compiler+runtime/src/cpp/jank/runtime/obj/integer_range.cpp

```diff
@@ -0,0 +1,193 @@
+
+namespace jank::runtime
+{
+  static native_bool positive_step_bounds_check(native_integer const val, native_integer const end)
+  {
+    return end <= val;
+  }
+
+  static native_bool negative_step_bounds_check(native_integer const val, native_integer const end)
+  {
+    return val <= end;
+  }
+
+  obj::integer_range::static_object(native_integer const end)
+    : end{ end }
+    , step{ 1 }
+    , bounds_check{ positive_step_bounds_check }
+  {
+  }
+
+  obj::integer_range::static_object(native_integer const start, native_integer const end)
+    : start{ start }
+    , end{ end }
+    , step{ 1 }
+    , bounds_check{ positive_step_bounds_check }
+  {
+  }
+
+  obj::integer_range::static_object(native_integer const start,
+                                    native_integer const end,
+                                    native_integer const step)
+    : start{ start }
+    , end{ end }
+    , step{ step }
+    , bounds_check{ step > 0 ? positive_step_bounds_check : negative_step_bounds_check }
+  {
+  }
+
+  obj::integer_range::static_object(native_integer const start,
+                                    native_integer const end,
+                                    native_integer const step,
+                                    obj::integer_range::bounds_check_t const bounds_check)
+    : start{ start }
+    , end{ end }
+    , step{ step }
+    , bounds_check{ bounds_check }
+  {
+  }
+
+  object_ptr obj::integer_range::create(native_integer const end)
+  {
+    if(end)
+    {
+      return make_box<obj::integer_range>(0, end, 1, positive_step_bounds_check);
+    }
+    return obj::persistent_list::empty();
+  }
+
+  object_ptr obj::integer_range::create(native_integer const start, native_integer const end)
+  {
+    return create(start, end, 1);
+  }
+
+  object_ptr obj::integer_range::create(native_integer const start,
+                                        native_integer const end,
+                                        native_integer const step)
+  {
+    if((step > 0 && end < start) || (step < 0 && start < end) || start == end)
+    {
+      return obj::persistent_list::empty();
+    }
+    /* TODO: Repeat object. */
+    //else if(is_zero(step))
+    //{ return make_box<obj::repeat>(start); }
+    return make_box<obj::integer_range>(start,
+                                        end,
+                                        step,
+                                        step > 0 ? positive_step_bounds_check
+                                                 : negative_step_bounds_check);
+  }
+
+  obj::integer_range_ptr obj::integer_range::seq()
+  {
+    return this;
+  }
+
+  obj::integer_range_ptr obj::integer_range::fresh_seq() const
+  {
+    return make_box<obj::integer_range>(start, end, step, bounds_check);
+  }
+
+  object_ptr obj::integer_range::first() const
+  {
+    return make_box<obj::integer>(start);
+  }
+
+  obj::integer_range_ptr obj::integer_range::next() const
+  {
+    if(count() < 1)
```

Comment:
What you want to check is if `count() == 0`. Why are we checking if it's `< 1` instead? Same effect, but much less clear, since the implication here might be that it could be less than zero (which it cannot be, given that it's an unsigned type).

Reply:
Sorry, I think both should be count() <= 1, because a range returns a less than or equal to range, so (range 3)  = (0 1 2). If this was == 0, it would return (0, 1, 2, 3).